changelog/deprecate-synchronized.dd

```diff
@@ -0,0 +1,28 @@
+Add `-preview=nosynchronized`: Deprecate plain `synchronized` statements.
+
+Synchronized statements have two forms, `synchronized` and `synchronized(exp)`.
+When there is no expression argument, a global mutex is created using a buffer
+big enough to store the platform dependant critical section type.
+
+Example:
+```
+void main()
+{
+    synchronized {
+        // __gshared byte[40] __critsec;
+        // _d_criticalenter(&__critsec[0]);
+        // scope(exit) _d_criticalexit(&__critsec[0]);
+        (cast() counter) += 1;
+    }
+}
+```
+
+This implementation suffers from a couple deficiencies. Neither the size nor
+alignment of the `critsec` buffer are obtained from the OS critical section
+type defined in druntime. As a result, if the size allocated by the compiler is
+too small, or the platform has strict alignment requirements, then the program
+run-time will crash on entering the synchronized statement.
+
+With `-preview=nosynchronized`, this code will now issue a deprecation.
+To migrate away, you can either provide an explicit mutex argument for
+`synchronized`, or use the `core.sync.mutex.Mutex` library type in druntime.
```

Comment:
Or the compiler can simply internally replace `synchronized()` with `synchronized(core.sync.mutex.Mutex)`